### DISCUSSION TITLE
MT7921, BCM4350, QCA6174, Intel 8260 - Test Results

### DISCUSSION THREAD

**Discussion by thinkingsand** [vanilla-wiiu/vanilla#17] [Hardware Results]
As discussed in the discord, posting it all here for clarity. All tested with the `6faf37a` action build. I also tested the MT7921 on a version compiled from scratch, didn't seem to make a difference.

## MediaTek MT7921
Modern WiFi 6 card - tested on an AMD-based ROG Strix G15 2021 running Arch and Fedora 39. Same result on both.
Pairing, connecting to the console and input all worked. Audio and video output did not. No errors, the output screen just stays blank.

## Broadcom BCM4350
Used in the Dell DW1820A card popular in Hackintosh builds. Tested on an Intel-based HP EliteBook 850 G5 running Ubuntu 24.04.  
Pairing worked. Connecting back to the console didn't - it looped endlessly on `INFO WAITING FOR CONNECTION`. Similar failure mode to the other Broadcom chips, seems they're all like this.

## Qualcomm QCA6174
Tested on a 1st gen Surface Go running Fedora 40.  
Same issues as the MT7921, input works but audio and video don't.

## Intel Wireless-AC 8260
Finally, success! Works fine, all features seem to work.  
Tested on an Intel-based HP EliteBook 840 G3 running Ubuntu 24.04

**Comment by itsmattkc** [comment]
I've had a realization - on the setups where **input** works but **video/audio** don't, do you have a firewall installed?

Wii U gamepad communication is performed over a handful of UDP ports and obviously controller input is an outbound connection while video and audio are inbound, so if a firewall is blocking inbound connections on those ports, it won't be able to receive anything.

**Comment by thinkingsand** [comment]
Just retested and yup, turns out that was the issue all along. 
I'll update the wiki as the MediaTek card is now confirmed working. Probably worth adding a quick note to the documentation, I imagine it'll trip a lot of people up.

The Qualcomm chip is giving me the same result as #20, with working input and audio but no video, and a crash after around 30 seconds.

**Comment by thinkingsand** [comment]
Sorry for the edit, mixed up laptops the first time round and tested the wrong one.

**Comment by itsmattkc** [comment]
Great, thanks for re-testing that for me, I'll definitely make a note. I wonder if there's a safe way to configure the firewall to port forward Vanilla? Might be more effort than it's worth so for the time being we'll just advise people to manually configure/disable their firewall.

Qualcomm card is interesting, there's [this point](https://youtu.be/N6sY8-yjRRk?t=246) in one of my videos where I mention I could "never quite get drc-sim to work properly". I didn't elaborate, but the issue was I could control the console and get audio, but got no video. This was with the Surface Laptop 3 AMD version, whose Wi-Fi chipset is a QCA61x4A; very similar model number to the two cards that exhibit that issue. Might indeed be a hardware incompatibility plaguing that whole series (unfortunately I don't have that laptop anymore so I can't re-test it with Vanilla).

**Reply by Paulinhoice** [reply]
What can we do to help you identify the root cause of the QCA board bug?

**Reply by thinkingsand** [reply]
There's nothing obvious going wrong - nothing in the vanilla/kernel log. I'll take another look tomorrow and see if the inbound video connection is being opened, and possibly take a packet capture.

**Reply by itsmattkc** [reply]
It is curious, you'd think if it could receive one kind of packet (audio), it could receive the others too.

I do know video frames are sent in segments - I think 4-5 separate segments per frame - and if any one of those isn't received, the frame can't be reconstructed. Perhaps one of those gets lost in transit and messes things up.

**Reply by spicyegg234** [reply]
> Qualcomm card is interesting, there's [this point](https://youtu.be/N6sY8-yjRRk?t=246) in one of my videos where I mention I could "never quite get drc-sim to work properly". I didn't elaborate, but the issue was I could control the console and get audio, but got no video. This was with the Surface Laptop 3 AMD version, whose Wi-Fi chipset is a QCA61x4A; very similar model number to the two cards that exhibit that issue. Might indeed be a hardware incompatibility plaguing that whole series (unfortunately I don't have that laptop anymore so I can't re-test it with Vanilla).



> It is curious, you'd think if it could receive one kind of packet (audio), it could receive the others too.
> 
> I do know video frames are sent in segments - I think 4-5 separate segments per frame - and if any one of those isn't received, the frame can't be reconstructed. Perhaps one of those gets lost in transit and messes things up.


@itsmattkc bad MTU? could be a driver/firmware limitation on Qualcomm cards, though as far as i can tell its patchable since the hardware should follow 802.11 (basically repeating what i said below)

**Reply by itsmattkc** [reply]
I would strongly suspect MTU issue, or something similar (perhaps limited UDP receive buffer). Video packets are the largest (usually approaching 1,800 bytes), followed by audio packets (usually ~1,664 bytes). Input packets are only 128 bytes so they have no issue getting through. If for whatever reason the chip rejected packets over a certain size, that would explain all of these symptoms.

(since the issue is size-related, a very small compressed video frame could get through, i.e. one with minimal complexity, like a blank white screen, but the Wii U doesn't have many of those).

**Reply by spicyegg234** [reply]
clearly we need a drh-sim (to simulate the wii u end) with an adjustable MTU for troubleshooting! basically just a router with modified WPA2 and WPS and sending a h264 stream and pcm audio (im fairly sure standard wireless hardware can be a router). would actually be pretty cool to use a real gamepad on. might be out of scope for Vanilla though... perhaps modifying libdrc for a quick and dirty hack for testing? my Dell Precision 3520 has an Qualcomm/Atheros card that does use ath10k, so i can try messing with that

**Reply by spicyegg234** [reply]
oh, my Precision has the QCA6174. huh, well perfect!
even two years later, i can confirm this card can pair and get DHCP established, but doesnt get video. I havent tested input yet, and at first it doesnt seem to get audio either. but after sniffing around with tcpdump, i saw consistent ~1672 byte packets on UDP port 50121. More importantly, not only is this greater than a MTU of 1500, capturing these packets and stripping the header, then shoving it into aplay, i did indeed get the wii u's audio stream. however, for the video packets (which seem to be smaller at 1100/1200 bytes!, to be fair its just the aroma boot menu with the blue background) it seemed to get most of them? but enough packets seemed to be consistently dropped somehow, perhaps the 1800 byte packets? i would have to compare with a packet dump from my Intel AC 8265 on my Latitude to see what is "supposed" to be coming through

**Reply by ashman-777** [reply]
> oh, my Precision has the QCA6174. huh, well perfect! even two years later, i can confirm this card can pair and get DHCP established, but doesnt get video. I havent tested input yet, and at first it doesnt seem to get audio either. but after sniffing around with tcpdump, i saw consistent ~1672 byte packets on UDP port 50121. More importantly, not only is this greater than a MTU of 1500, capturing these packets and stripping the header, then shoving it into aplay, i did indeed get the wii u's audio stream. however, for the video packets (which seem to be smaller at 1100/1200 bytes!, to be fair its just the aroma boot menu with the blue background) it seemed to get most of them? but enough packets seemed to be consistently dropped somehow, perhaps the 1800 byte packets? i would have to compare with a packet dump from my Intel AC 8265 on my Latitude to see what is "supposed" to be coming through

The diagnosis I got back from Claude was that packet dropping (now I think suspected to be  issue in ath10k / mtu ) then triggers an issue in video.c logic IDR further down managing packet drops etc. The second thing (if it really is a thing) won’t fix the video as I think the packets are already messed up so garbage in garbage out. Hence I think how I managed to get a garbled screen out and initially you can see a few clear screens.

On Ath10k I only got as far as reading up on HTT_RX_BUF_SIZE 1920 but didn’t work out much else. As it was not straightforward, at least with my skills , to see if changing that would cause any change in behaviour of the video feed.

Edit: comparison of connections usb WiFi vs qca6174

| Adapter | Chipset/Driver | Signal | TX Rate | RX Rate | Result |
|---|---|---|---|---|---|
| USB WiFi linksys ae1000 | VHT-capable (unspecified) | -43 dBm | 162.0 Mbit/s (VHT-MCS 4, 40MHz, NSS2) | 866.7 Mbit/s (VHT-MCS 9, 80MHz, short GI) | ok works fine |
| Internal PCIe | QCA6174 rev 32 (ath10k) | -54 dBm | 130.0 Mbit/s (MCS 15) | 58.5 Mbit/s (MCS 6) | Troubleshooting |

**Reply by spicyegg234** [reply]
some more testing, here is what i found out:

on port 50120 (video stream):

|Adapter|Input|Audio|Video|Max Packet Size Seen|Conclusion|
|---|---|---|---|---|---|
|Qualcomm QCA6174|✅ Working| ⚠️ Working*| ❌ Not Working | ~1675 bytes | Something definitely fishy is going on, the biggest of packets are being dropped (like the I frames needed to start a h264 stream...).<br><br>*This also seems to affect the audio stream, as it seems to sound fairly choppy and stuttery, but since its just raw PCM, it can get away with packets being lobotomized.
|Intel AC 8265|✅ Working|✅ Working|✅ Working| ~1790 Bytes | Works perfectly fine |

**Reply by ashman-777** [reply]
> some more testing, here is what i found out:
> 
> on port 50120 (video stream):
> 
> Adapter	Input	Audio	Video	Max Packet Size Seen	Conclusion
> Qualcomm QCA6174	✅ Working	⚠️ Working*	❌ Not Working	~1675 bytes	Something definitely fishy is going on, the biggest of packets are being dropped (like the I frames needed to start a h264 stream...).
> 
> *This also seems to affect the audio stream, as it seems to sound fairly choppy and stuttery, but since its just raw PCM, it can get away with packets being lobotomized.
> Intel AC 8265	✅ Working	✅ Working	✅ Working	~1790 Bytes	Works perfectly fine

Does this command work for you to output log file? I’m interested if you get loads of damn, IDR log messages. I was only able to get touch controls after I changed video.c how it recovered from packet losses.

./vanilla > ~/vanilla.log 2>&1

**Reply by itsmattkc** [reply]
Also does anything change if you force a higher MTU:

```
sudo ip link set dev wlan0 mtu 1800
```

(change `wlan0` to the correct interface)

**Reply by ashman-777** [reply]
> Also does anything change if you force a higher MTU:
> 
> ```
> sudo ip link set dev wlan0 mtu 1800
> ```
> 
> (change `wlan0` to the correct interface)

Thanks. No change for me.

 I’m testing on Mario Kart 8 as if you leave it at start screen it loops through from a black loading screen, then to gameplay demo, then to grey screen etc. The music track plays over the whole lot on loop video transitions so you can hear the choppy audio changes. Video and Audio is fine on black and grey screens where there is no video. Touch inputs all fine. 
 
So I think as all have suggested previously it’s up in the chain. My video.c does have a change to how it responds where packets get dropped. In the original video.c I just get stuck on the connection screen - I just see lots of damn, IDR errors in the log.

**Reply by spicyegg234** [reply]
> Also does anything change if you force a higher MTU:
> 
> ```
> sudo ip link set dev wlan0 mtu 1800
> ```
> 
> (change `wlan0` to the correct interface)


no, not at all. i dont think its a MTU issue anymore, atleast not hardware. when i set that it A: didnt complain and B: the network still worked. im hopeful its fixable in software/firmware

i can try making a router have a MTU of 1800 to try and recreate the conditions of the wii u


> 
> Does this command work for you to output log file? I’m interested if you get loads of damn, IDR log messages. I was only able to get touch controls after I changed video.c how it recovered from packet losses.
> 
> ./vanilla > ~/vanilla.log 2>&1

oh, no need! I can confidently tell you it spammed the fuck out of IDR errors (and decoder errors once it did technically start the stream)

**Reply by itsmattkc** [reply]
Here's something interesting, a 2020 Linux kernel patch to the ath10k driver (the driver that should be in use here) explicitly claims that it "increased the rx buffer size" such that the maximum frame size was 1656 bytes (up from 1528): https://lists.openwall.net/netdev/2020/02/05/109

This isn't related to the MTU setting, it's a whole separate limitation on incoming frame size, and sounds pretty close to the figure you found in your testing. I suspect that might be the reason the larger packets are getting filtered out. If so, it may be possible to increase this buffer size in the driver and recompile it, but that also might break things (presumably the limit exists for a reason, maybe a hardware/firmware limitation?). Still, might be worth a try.

Also worth noting that the WCN3990 mentioned in #481 would also be covered under the same ath10k driver, which would likely explain the issue there too.

**Reply by spicyegg234** [reply]
> Here's something interesting, a 2020 Linux kernel patch to the ath10k driver (the driver that should be in use here) explicitly claims that it "increased the rx buffer size" such that the maximum frame size was 1656 bytes (up from 1528): https://lists.openwall.net/netdev/2020/02/05/109
> 
> This isn't related to the MTU setting, it's a whole separate limitation on incoming frame size, and sounds pretty close to the figure you found in your testing. I suspect that might be the reason the larger packets are getting filtered out. If so, it may be possible to increase this buffer size in the driver and recompile it, but that also might break things (presumably the limit exists for a reason, maybe a hardware/firmware limitation?). Still, might be worth a try.
> 

i can try fiddling with the driver and seeing if more rx ring buffer = more fixed

> Also worth noting that the WCN3990 mentioned in #481 would also be covered under the same ath10k driver, which would likely explain the issue there too.

thats also what i was thinking, afaik even the mobile chipset is softmac (it did get DHCP established)

**Reply by ashman-777** [reply]
> Here's something interesting, a 2020 Linux kernel patch to the ath10k driver (the driver that should be in use here) explicitly claims that it "increased the rx buffer size" such that the maximum frame size was 1656 bytes (up from 1528): https://lists.openwall.net/netdev/2020/02/05/109
> 
> This isn't related to the MTU setting, it's a whole separate limitation on incoming frame size, and sounds pretty close to the figure you found in your testing. I suspect that might be the reason the larger packets are getting filtered out. If so, it may be possible to increase this buffer size in the driver and recompile it, but that also might break things (presumably the limit exists for a reason, maybe a hardware/firmware limitation?). Still, might be worth a try.
> 
> Also worth noting that the WCN3990 mentioned in #481 would also be covered under the same ath10k driver, which would likely explain the issue there too.

Thanks. I was also looking at the ath10k-ct but I haven’t figured out how to install it and if it would work with QCA6174.

**Comment by RobotZGalaxy** [comment]
How long did syncing take for you with the MediaTek MT7921? Upon attempting it just gets stuck for me.

**Reply by bitsandbytes0** [reply]
wish someone would respond to this

**Reply by EMREOYUN** [reply]
> wish someone would respond to this

For my case, different times between from 15 seconds up to 4 minutes. You can enable logs for troubleshooting.

**Comment by EMREOYUN** [comment]
MediaTek MT7921 on Asus Tuf F15 working without issues on Ubuntu 24 but it required an ethernet connection on my end.

**Reply by bitsandbytes0** [reply]
Isn’t Ethernet a different controller?

**Reply by EMREOYUN** [reply]
> Isn’t Ethernet a different controller?

There is somehow an unreplicated bug for controllers that happened to me but it only occurred on Ubuntu 24. Testing on other Debian based OSes didn't have this issue.

**Comment by borosf** [comment]
Weird, I have a Dell Latitude running CachyOS with a QCA6174. The laptop seems to have synced with the Wii U but I am stuck on "Connecting to Wii U".

**Reply by ashman-777** [reply]
> Wii

I have the QCA6174 built into the Acer Chromebook tab 10. Similar experience in that the screen gets stuck on “Connecting to Wii U” with a black screen displaying “ sending IDR incomplete frame”. Interesting though in clicking on systems settings which from memory was gamepad only kicked the system to start up doing something - a grey screen and audio (choppy) with a log file change.

I’m not a programmer (and I have a working usb WiFi) but I like this challenge of partial working (it’s cleaner not having a usb hub and adapter attached). Vanilla builds out in around a minute so hoping I can learn a bit of C , ffmpeg along the way. 

<img width="3024" height="3023" alt="IMG_7498" src="https://github.com/user-attachments/assets/1b7ddd97-2c24-4c53-b080-f5f150dc5d88" />
<img width="3024" height="2913" alt="IMG_7493" src="https://github.com/user-attachments/assets/0d015a6f-fa4f-408d-8005-5bcd34f482af" />
<img width="3024" height="4032" alt="IMG_7489" src="https://github.com/user-attachments/assets/71eb354b-c716-425a-9801-ee810817c321" />

**Comment by spicyegg234** [comment]
@ashman-777 i suspect the severe corruption and video glitching is because the card cant do a MTU of 1800, whether is a software (driver), firmware, or hardware limitation i dunno. you should try setting a higher MTU than 1500 and see if the card/driver accepts it.

best case scenario: driver limitation, just patch the C source
probable scenario: firmware limitation, need to patch a undocumented binary blob
worst case scenario: hardware does not follow 802.11 spec and cannot do 1800 MTU (rare)
> i meant to make a reply oops

**Reply by ashman-777** [reply]
> @ashman-777 i suspect the severe corruption and video glitching is because the card cant do a MTU of 1800, whether is a software (driver), firmware, or hardware limitation i dunno. you should try setting a higher MTU than 1500 and see if the card/driver accepts it.
> 
> best case scenario: driver limitation, just patch the C source probable scenario: firmware limitation, need to patch a undocumented binary blob worst case scenario: hardware does not follow 802.11 spec and cannot do 1800 MTU (rare)
> 
> > i meant to make a reply oops

Thanks. I had a little play around with mtu, a bit with the vanilla code, and an earlier ath10k firmware but not able to change behaviour. I do see there is some work on this ath10k so there might be a fixed firmware out there.

**Reply by ashman-777** [reply]
On the QCA6174 I’ve messed around with video.c code (combination of myself bouncing ideas off Claude Ai) and got it to this weird state where in the pause menu I can “wipe the screen clean” using the cursor. Still ‘partially working’ and I need to go through the log as still lots of corruption but seemed to improve a little. I’m thinking it maybe a cpu loading issue so might look at dropping the terminal logging.

<img width="3024" height="2163" alt="IMG_7519" src="https://github.com/user-attachments/assets/b9bf6a80-04c3-4044-846e-48f611c18d0c" />
<img width="2997" height="2652" alt="IMG_7517" src="https://github.com/user-attachments/assets/873ae137-811b-4588-bf99-efdbd77115c9" />

**Reply by borosf** [reply]
If more testing is needed, give me a shout. I still have this laptop.
